Add skip indices to query log

grantholly-clickhouse opened the pull request:
<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):

- Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

We now track any data skipping indices that were used during query execution in the query_log table in a new column named `skip_indices`. Fixes https://github.com/ClickHouse/ClickHouse/issues/78676. Original author @pheepa

### Documentation entry for user-facing changes

Adds a new Array[String] column to system.query_log for any data skipping indexes used during query execution

### Details

Here I am trying to add the list of skip indexes used during execution to the query_log table. I want to add the skip index to the QueryAccessInfo if we actually used the skip index to further filter down granules. I don't want to add the skip index if it is only considered during planning. The new query_log table should have an array column where we keep all the skip indexes that were used during the query. I'm grabbing the index name from the index condition type, and I hope, because I'm not sure, that this should match the name we use for the data_skipping_indexes.name column.

I borrowed heavily from the way that we track projections and views, hence the new method, and array type column.

My hope is that this will provide another way to troubleshoot ClickHouse queries and index utilization.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Adds a new `system.query_log` column and populates it from MergeTree skip-index filtering, which changes query log schema and touches MergeTree index analysis paths. Risk is mostly around correctness/compatibility of log output and potential overhead when collecting index names.
> 
> **Overview**
> Adds a new `skip_indices` Array(LowCardinality(String)) column to `system.query_log` and documents it in `query_log.md`.
> 
> Plumbs skip-index usage through `Context::QueryAccessInfo` and `QueryLogElement`, and records fully-qualified skip index names when MergeTree skip indexes are actually used to filter ranges (`MergeTreeDataSelectExecutor`). Includes a new stateless test validating `skip_indices` output for simple filters and JOIN scenarios.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b235a3bb3eceee65abd8837438e8d27eba93804a. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

grantholly-clickhouse committed:
adding initial array column to the query_log table to store skip indexes used during execution

grantholly-clickhouse committed:
adding skip index tracking to query execution and QueryAccessInfo

grantholly-clickhouse committed:
tracking selected skip indexes

grantholly-clickhouse committed:
attempting to add used skip indexes to QueryAccessInfo

grantholly-clickhouse committed:
correctly getting modifiable QueryAccessInfo

grantholly-clickhouse committed:
factoring out QueryAccessInfo method into a new method addSkipIndexAccessInfo in the Context file

grantholly-clickhouse committed:
adding query test and reference files

grantholly-clickhouse committed:
recording skip index usage without check and updating docs

grantholly-clickhouse committed:
fixing style issue

grantholly-clickhouse committed:
fixing an unintentional change

grantholly-clickhouse committed:
Merge remote-tracking branch 'origin/master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

azat commented on tests/queries/0_stateless/03632_skip_indexes_logged_in_query_log.sql:
```diff
@@ -0,0 +1,55 @@
+
+SYSTEM FLUSH LOGS system.query_log;
+
+SELECT arraySort(skip_indexes)
```
Let's also add `query` - so that .reference file make sense w/o .sql

```suggestion
SELECT formatQuerySingleLine(query), skip_indexes
```

grantholly-clickhouse replied:
The only thing that makes me hesitate to do this is that, if we use `database.table.index_name`, we would need some way to know the name of the database during the test.  Otherwise I'm not sure how we would have a predictable reference file output.

azat replied:
No need to, `clickhouse-test` replaces randomly generated database name with `default`

azat replied:
And please also add a `query` for easier review of `reference`

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Update docs/en/operations/system-tables/query_log.md

changing indexes to indices

Co-authored-by: Azat Khuzhin <a3at.mail@gmail.com>

grantholly-clickhouse committed:
adding table name to skip index

grantholly-clickhouse commented on src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:
```diff
@@ -878,6 +878,8 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
                             log);
                     }
 
+                    context->getQueryContext()->addSkipIndexAccessInfo(index_and_condition.index->index.name);
```
Here's what  I have now 
```
SELECT skip_indexes
FROM system.query_log
WHERE query_id = '172fe960-387d-4d54-9d10-23fa1a22fd33'

Query id: 8bc2f5d3-cd9c-475d-acb2-47ab655b0adc

   ┌─skip_indexes──────────┐
1. │ ['test.users.name_i'] │
2. │ ['test.users.name_i'] │
   └───────────────────────┘
```

grantholly-clickhouse committed:
fixing types in documentation for the query_log table

azat commented on src/Interpreters/Context.cpp:
```diff
@@ -2364,6 +2364,15 @@ void Context::addQueryAccessInfo(const QualifiedProjectionName & qualified_proje
         "{}.{}", qualified_projection_name.storage_id.getFullTableName(), backQuoteIfNeed(qualified_projection_name.projection_name)));
 }
 
+void Context::addSkipIndexAccessInfo(const String & table_name, const String & skip_index_name)
+{
+    if (skip_index_name.empty())
+        return;
+
+    std::lock_guard lock(query_access_info->mutex);
+    query_access_info->skip_indexes.emplace(table_name + "." + skip_index_name);
+}
```
```suggestion
void Context::addSkipIndexAccessInfo(const String & full_table_name, const String & skip_index_name)
{
    if (skip_index_name.empty())
        return;

    std::lock_guard lock(query_access_info->mutex);
    query_access_info->skip_indexes.emplace(full_table_name + "." + skip_index_name);
}
```

azat commented on tests/queries/0_stateless/03632_skip_indexes_logged_in_query_log.reference:
```diff
@@ -0,0 +1,5 @@
+['age_i']
+['name_i']
+['age_i','name_i']
+['age_i','name_i']
```
You forgot to update the reference.

azat commented on docs/en/operations/system-tables/query_log.md:
```diff
@@ -69,6 +69,7 @@ You can use the [log_formatted_queries](/operations/settings/settings#log_format
 - `columns` ([Array](../../sql-reference/data-types/array.md)([LowCardinality(String)](../../sql-reference/data-types/lowcardinality.md))) — Names of the columns present in the query.
 - `partitions` ([Array](../../sql-reference/data-types/array.md)([LowCardinality(String)](../../sql-reference/data-types/lowcardinality.md))) — Names of the partitions present in the query.
 - `projections` ([String](../../sql-reference/data-types/string.md)) — Names of the projections used during the query execution.
+- `skip_indexes` ([String](../../sql-reference/data-types/string.md)) — Names of any data skipping indexes that were used during query execution
```
You forgot to rename the column.

azat commented on src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:
```diff
@@ -878,6 +878,11 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
+                        ranges.data_part->storage.getStorageID().getFullTableName(),
+                        index_and_condition.index->index.name
+                    );
+
```
Let's also add the same for `merged_indices` below, and update test. This is a `hypothesis` index, you can take a look at i.e. `tests/queries/0_stateless/01624_soft_constraints.sh` for example

grantholly-clickhouse committed:
Update src/Interpreters/Context.cpp

making argument name more descriptive

Co-authored-by: Azat Khuzhin <a3at.mail@gmail.com>

grantholly-clickhouse committed:
updating query_log docs, making argument names more descriptive, and updating SQL query test

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
updating reference file

grantholly-clickhouse committed:
removing + characters copied from output

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
renaming indexes to indices

grantholly-clickhouse committed:
updating tests to use new column name

grantholly-clickhouse committed:
fixing call to QueryAccessInfo

grantholly-clickhouse committed:
adding skip index tracking for merged indexes

grantholly-clickhouse committed:
capturing all index names in merged index

grantholly-clickhouse committed:
renaming test files from 'indexes' to 'indices'

grantholly-clickhouse committed:
updating test to check for merged indices

grantholly-clickhouse committed:
updating reference file

grantholly-clickhouse committed:
removing newline at end of reference file

grantholly-clickhouse committed:
cleaning up test files

grantholly-clickhouse committed:
style fixes

grantholly-clickhouse committed:
removing extra whitespace

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
updating reference file

grantholly-clickhouse committed:
updating reference file

grantholly-clickhouse committed:
updating reference file

grantholly-clickhouse committed:
updating reference file

grantholly-clickhouse committed:
updating reference file

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
updating reference file

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
adding no-parallel-replicas tag to test

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log

grantholly-clickhouse committed:
updating code to no longer work with hypothesis indexes after merging in upstream

azat commented on src/Interpreters/Context.cpp:
```diff
@@ -2554,6 +2554,15 @@ void Context::addQueryAccessInfo(const QualifiedProjectionName & qualified_proje
 
+void Context::addSkipIndexAccessInfo(const String & full_table_name, const String & skip_index_name)
+{
+    if (skip_index_name.empty())
```
How it can be empty?

grantholly-clickhouse replied:
I can't remember why I put this in earlier.  

I think we can remove it.  I don't think you could have a skip index in the database that had an empty name.

azat reviewed (commented):
In general looks ok, I've adjusted title and description and wrote couple of comments, PTAL

Also, it will be nice to support amount of filtered out granules, but this has some complexity due to recently added `use_skip_indexes_on_data_read` (though it should not be hard to do), what do you think?

Also, maybe we need to add `Co-authored-by` for the author from https://github.com/ClickHouse/ClickHouse/pull/79635, to give him credits?

grantholly-clickhouse committed:
Update src/Interpreters/Context.h

Co-authored-by: Azat Khuzhin <a3at.mail@gmail.com>

grantholly-clickhouse committed:
Moving calls to addSkipIndexAccessInfo out of the loop.  We now call it
after `useful_indices` has been built.  I've added some checks make sure
we have a valid query context

grantholly-clickhouse committed:
Update src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

Co-authored-by: Azat Khuzhin <a3at.mail@gmail.com>

grantholly-clickhouse committed:
adding StorageID field to IndexAnalysisContext

grantholly-clickhouse committed:
changing new IndexAnalyssiContext field to store value instead of const ref

alexbakharew commented:
Hi @grantholly-clickhouse,
could you please merge master into your branch? It will resolve LLVM Coverage issue

grantholly-clickhouse committed:
Merge branch 'master' into add_skip_indexes_to_query_log
